File: api/models/comment.py
from beanie import Document
from pydantic import Field
from datetime import datetime, timezone
from typing import Optional

class Comment(Document):
    author_id: str
    post_id: str
    parent_reply_id: Optional[str] = None # null = top-level comment
    
    content: str
    x_coordinate: Optional[float] = None # Used for mapping top-level comments on the image
    y_coordinate: Optional[float] = None
    
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    
    likes_count: int = 0
    replies_count: int = 0 # Denormalized reply count

    class Settings:
        name = "comments"
        indexes = [
            "post_id",
            "parent_reply_id"
        ]

File: ios/YapHub/Models/Comment.swift
import Foundation

nonisolated struct Comment: Codable, Identifiable, Sendable {
    let id: String
    let authorId: String
    let postId: String
    let parentReplyId: String?
    let content: String
    let xCoordinate: Double?
    let yCoordinate: Double?
    let createdAt: String
    let updatedAt: String
    var likesCount: Int
    var repliesCount: Int
    let author: Author?
    var isLiked: Bool?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case authorId = "author_id"
        case postId = "post_id"
        case parentReplyId = "parent_reply_id"
        case content
        case xCoordinate = "x_coordinate"
        case yCoordinate = "y_coordinate"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case likesCount = "likes_count"
        case repliesCount = "replies_count"
        case author
        case isLiked = "is_liked"
    }
}

nonisolated struct CreateCommentResponse: Codable, Sendable {
    let commentId: String

    enum CodingKeys: String, CodingKey {
        case commentId = "comment_id"
    }
}

nonisolated struct CreateReplyResponse: Codable, Sendable {
    let replyId: String

    enum CodingKeys: String, CodingKey {
        case replyId = "reply_id"
    }
}
